src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php

```diff
@@ -161,46 +156,33 @@ protected function doDelete(array $ids, array $tagData = []): bool
      */
     protected function doInvalidate(array $tagIds): bool
     {
-        if (!$this->redisServerSupportSPOP()) {
-            return false;
-        }
-
-        // Pop all tag info at once to avoid race conditions
-        $tagIdSets = $this->pipeline(static function () use ($tagIds) {
-            foreach ($tagIds as $tagId) {
-                // Client: Predis or PHP Redis 3.1.3+ (https://github.com/phpredis/phpredis/commit/d2e203a6)
-                // Server: Redis 3.2 or higher (https://redis.io/commands/spop)
-                yield 'sPop' => [$tagId, self::POP_MAX_LIMIT];
+        // Rename all sets so we can safely avoid race conditions
+        $this->pipeline(static function () use ($tagIds) {
+            $uniqueToken = bin2hex(random_bytes(10));
+            foreach ($tagIds as $key => $tagId) {
+                $tagIds[$key] = $temporaryTagId = $tagId . ':' . $uniqueToken;
+                yield 'rename' => [$tagId, $temporaryTagId];
             }
         });
 
-        // Flatten generator result from pipeline, ignore keys (tag ids)
-        $ids = array_unique(array_merge(...iterator_to_array($tagIdSets, false)));
+
+        // Now safely take the time to scan the keys in each set and collect ids we need to delete
+        $idHash = [];
+        foreach ($tagIds as $temporaryTagId) {
+            $cursor = null;
+            while(($ids = $this->redis->sScan($temporaryTagId, $cursor, null, self::SSCAN_LIMIT)) !== false) {
+                foreach ($ids as $id) {
+                    $idHash[$id] = true;
+                }
+            }
+            $idHash[$temporaryTagId] = true;
+        }
 
         // Delete cache in chunks to avoid overloading the connection
-        foreach (array_chunk($ids, self::BULK_DELETE_LIMIT) as $chunkIds) {
+        foreach (array_chunk(array_keys($idHash), self::BULK_DELETE_LIMIT) as $chunkIds) {
```

Comment:
You can put the ids as values and get rid of the array_keys calls, isn't it?

Reply:
👍 I'll change it.

Thinking was that if you clear several tags you'll have several duplicates of values, so to avoid array growing to large before calling array_unque. But each an every set won't contain duplicates as guaranteed by Set datatype, so amount of duplicates will be negligible.